File: services/pix-boleto-kotlin/src/main/kotlin/com/acme/payments/pixboleto/messaging/PaymentProcessingCommandMapper.kt
package com.acme.payments.pixboleto.messaging

import com.acme.payments.pixboleto.domain.ChargeCommand
import com.acme.payments.pixboleto.domain.PaymentRail
import com.acme.payments.pixboleto.provider.PaymentProvider
import org.springframework.stereotype.Component
import java.time.LocalDate

@Component
class PaymentProcessingCommandMapper {
    fun toChargeCommand(event: PaymentProcessingEvent): ChargeCommand? {
        val rail = when (event.method) {
            PaymentRail.PIX.name -> PaymentRail.PIX
            PaymentRail.BOLETO.name -> PaymentRail.BOLETO
            else -> return null
        }

        return ChargeCommand(
            idempotencyKey = event.idempotencyKey,
            rail = rail,
            amount = event.amount,
            currency = event.currency.uppercase(),
            dueDate = dueDate(event),
            preferredProvider = provider(event),
        )
    }

    private fun provider(event: PaymentProcessingEvent): PaymentProvider? {
        val provider = event.provider ?: event.metadata["preferredProvider"]?.toString()
        return provider?.let(PaymentProvider::valueOf)
    }

    private fun dueDate(event: PaymentProcessingEvent): LocalDate? {
        return event.metadata["dueDate"]?.toString()?.let(LocalDate::parse)
    }
}

File: services/card-payment-csharp/src/CardPaymentService/Messaging/PaymentProcessingCommandMapper.cs
using System.Text.Json;
using Acme.Payments.CardPaymentService.Domain;

namespace Acme.Payments.CardPaymentService.Messaging;

public sealed class PaymentProcessingCommandMapper
{
    public CardPaymentCommand? ToCommand(PaymentProcessingEvent paymentEvent)
    {
        if (!string.Equals(paymentEvent.EventType, RabbitPaymentTopology.PaymentProcessingRoutingKey, StringComparison.Ordinal) ||
            !string.Equals(paymentEvent.Method, "CREDIT_CARD", StringComparison.Ordinal))
        {
            return null;
        }

        return new CardPaymentCommand(
            paymentEvent.IdempotencyKey,
            paymentEvent.Amount,
            paymentEvent.Currency.ToUpperInvariant(),
            MetadataInt(paymentEvent, "installments") ?? 1,
            MetadataString(paymentEvent, "cardToken") ?? $"event-token:{paymentEvent.PaymentId}",
            Provider(paymentEvent)
        );
    }

    private static PaymentProvider? Provider(PaymentProcessingEvent paymentEvent)
    {
        string? provider = paymentEvent.Provider ?? MetadataString(paymentEvent, "preferredProvider");
        return provider?.ToUpperInvariant() switch
        {
            "ASAAS" => PaymentProvider.Asaas,
            "MERCADO_PAGO" => PaymentProvider.MercadoPago,
            "PAGBANK" => PaymentProvider.PagBank,
            "IUGU" => PaymentProvider.Iugu,
            "STRIPE" => PaymentProvider.Stripe,
            _ => null
        };
    }

    private static string? MetadataString(PaymentProcessingEvent paymentEvent, string key)
    {
        if (!paymentEvent.Metadata.TryGetValue(key, out JsonElement value))
        {
            return null;
        }

        return value.ValueKind switch
        {
            JsonValueKind.String => value.GetString(),
            JsonValueKind.Number => value.GetRawText(),
            _ => null
        };
    }

    private static int? MetadataInt(PaymentProcessingEvent paymentEvent, string key)
    {
        if (!paymentEvent.Metadata.TryGetValue(key, out JsonElement value))
        {
            return null;
        }

        if (value.ValueKind == JsonValueKind.Number && value.TryGetInt32(out int number))
        {
            return number;
        }

        return value.ValueKind == JsonValueKind.String && int.TryParse(value.GetString(), out number)
            ? number
            : null;
    }
}
